### PR TITLE
Notifications: Fix duplicate conflicting stores

### PR DESCRIPTION
**Pull request by kevin940726**
Related to #93530, https://github.com/Automattic/wp-calypso/pull/30603#discussion_r254538603.

## Proposed Changes

This PR unifies the notifications stores into one to prevent them from conflicting.

1. `init` is changed to a pure function which returns the initialized store.
2. Add a `createListenerMiddleware` to replace `actionMiddleware`. The new API allows adding listeners (handlers) dynamically without having to merge them beforehand. Some original usages are not migrated though to keep this PR small.
3. Close the global notifications when going to `/read/notifications` to avoid duplicate panels.
4. Don't render the global notifications panel when visiting `/read/notifications` to prevent conflicting listeners from registering at the same time.

## Why are these changes being made?

1. See https://github.com/Automattic/wp-calypso/issues/93530#issuecomment-2301347436 for the background issue and motivation. 
2. Some keyboard shortcuts are broken when there are multiple `<Notfications>` rendered on the page.
3. Visiting `/read/notifications` directly via the URL hangs on an infinite loading circle.

## Testing Instructions

Follow the original instructions in https://github.com/Automattic/wp-calypso/issues/93530.

1. Visit `/read/notifications` directly should not result in infinite loading circles.
2. Visit any other page on the reader, open the global notifications panel, and go to `/read/notifications` should close the global notifications panel.
3. Visit other sites by clicking the top left W admin button.
4. Hit <kbd>n</kbd> a few times and ensure the global notifications panel toggles.

## Pre-merge Checklist

- [x] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [x] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [x] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

### CODE REVIEW

**Review by Addison-Stavlo — COMMENTED** [review]
I love the idea.

But it seems like there might be an issue with the standalone build on this branch. If i test installing the standalone build on my sandbox and running that on simple sites outside of calypso, I cant seem to open the panel at all:

notes-panel-not-opening

**Comment by kevin940726** [comment]
> But it seems like there might be an issue with the standalone build on this branch.

Oops! Yep, there was an error in my code. I fixed that in https://github.com/Automattic/wp-calypso/pull/93766/commits/9ccb1ef01a80b33dc305413cb1f4c287d9a2c21d. I'm still figuring out the sandbox thing 😅. Please LMK if it works for you 🙇.

**Review by Addison-Stavlo — APPROVED** [review]
This works great on my end, both in calypso and on the standalone!

Thanks for the good improvements here!

**Comment by kevin940726** [comment]
> You can run the command from this comment above [#93766 (comment)](https://github.com/Automattic/wp-calypso/pull/93766#issuecomment-2301510434) with 'notifications' as the pluginSlug to get it on the sandbox.

Thanks! That's what I did too! There are surprisingly a lot of setups to make it work 😅. For some reasons, the dev build using oauth is not working for me though, and I'm not sure what went wrong there 🤔.

**Review by arthur791004 — APPROVED** [review]
Nice job!

File: apps/notifications/src/panel/state/index.js
Comment (jsnajdr) [inline apps/notifications/src/panel/state/index.js]: This could be just `create( reducer )`. The `createStore` function will construct the initial state itself.

File: apps/notifications/src/panel/Notifications.jsx
Comment (jsnajdr) [inline apps/notifications/src/panel/Notifications.jsx]: One thing that would remove a lot of confusion would be renaming `customMiddleware` to something like `actionHandlers`. Because it's not Redux middleware at all, it's a set of handlers for the `actionMiddleware`.

File: apps/notifications/src/panel/state/create-listener-middleware.ts
Comment (jsnajdr) [inline apps/notifications/src/panel/state/create-listener-middleware.ts]: This check is superfluous if you make sure that the listener middleware always runs after the thunk middleware. Which it does, because it's registered later.

File: apps/notifications/src/panel/state/create-listener-middleware.ts
Comment (jsnajdr) [inline apps/notifications/src/panel/state/create-listener-middleware.ts]: You can easily avoid the `!` operator by always doing the map lookup just once: instead of `.has` + `.get`, do just one `.get`.

File: apps/notifications/src/panel/state/index.js
Comment (jsnajdr) [inline apps/notifications/src/panel/state/index.js]: The `customMiddleware` parameter of `init` is no longer used, it's been replaced by adding listeners. So it can be removed. Also the `actionMiddleware` can be completely removed.

File: apps/notifications/src/panel/Notifications.jsx
Comment (jsnajdr) [inline apps/notifications/src/panel/Notifications.jsx:83]: This `initStore` call has any effect only in the `standalone` app, where the `customEnhancer` will assign the store to a local `store` variable. But the `initStore` function no longer assigns the store to the local `store` variable in the `state` module. It's now pure. But there are many modules, (`FilterBarController`, even the `Notifications` module I'm commenting on), that `import { store } from './state'`. They use the "wrong" store.

**Review by jsnajdr — COMMENTED** [review]
Thanks for trying to untangle the notifications Redux store. I think it's one of the first React/Redux apps ever created at Automattic, and it shows 🙂 I have many comments.

File: apps/notifications/src/panel/Notifications.jsx
Comment (jsnajdr) [reply apps/notifications/src/panel/Notifications.jsx:83]: One specific way how the `standalone` app will be failing now is when receiving messages from the parent window. The message handlers (`refresh`, `reset`) will dispatch actions to the local store, but the rest of the app is using the store in the `state` module.

A good solution would be to stop setting the local store, and just import `store` from `../panel/state`.

The end result is that all `Notifications` panels use the same instance of Redux store, but only add and remove their own listeners when mounting and unmounting.

File: apps/notifications/src/panel/state/create-listener-middleware.ts
Comment (kevin940726) [reply apps/notifications/src/panel/state/create-listener-middleware.ts]: I think it's a good practice to keep this independent from the setup, especially given that this is not a costly operation, WDYT?

File: apps/notifications/src/panel/state/create-listener-middleware.ts
Comment (kevin940726) [reply apps/notifications/src/panel/state/create-listener-middleware.ts]: I don't follow 🤔. Could you explain it further? 🙏

File: apps/notifications/src/panel/state/index.js
Comment (kevin940726) [reply apps/notifications/src/panel/state/index.js]: The `actionMiddleware` still has some default middlewares (handlers) set, so we still need it. We can migrate them to `listenerMiddleware` but I'd rather keep this PR small. We can certainly do that in another PR though!

Comment (jsnajdr) [reply apps/notifications/src/panel/state/index.js]: You're right, `actionMiddleware` still needs to be registered, we just can remove the `customMiddleware` parameter, and register it with a simple `actionMiddleware()` call.

Can be done in another PR so that this one doesn't get too complicated.

File: apps/notifications/src/panel/state/create-listener-middleware.ts
Comment (jsnajdr) [reply apps/notifications/src/panel/state/create-listener-middleware.ts]: The body of the loop can be rewritten as:
```js
let listenersForType = listeners.get( type );
if ( ! listenersForType ) {
  listenersForType = new Set< Handler >();
  listeners.set( type, listenersForType );
}
handlers.forEach( ( handler ) => listenersForType.add( handler ) );
```
This has two advantages:
- there's one `.get` call instead of `handlers.length + 1`
- TypeScript can figure all the types without any hints like `!`

File: apps/notifications/src/panel/state/create-listener-middleware.ts
Comment (jsnajdr) [reply apps/notifications/src/panel/state/create-listener-middleware.ts]: Yes, it's not a big deal, it's true that all Redux middlewares have a condition like "if the action is not my kind of thing, pass it throught to `next` immediately".

What's different here is that if the listener middleware encounters a non-plain action, it means that there is a bug: it's supposed to intercept all plain actions, but if registered in wrong order, it will miss the plain actions generated by other middlewares, like `redux-thunk`.

But it's no big deal, feel free to keep it as it is.

File: apps/notifications/src/panel/state/create-listener-middleware.ts
Comment (kevin940726) [reply apps/notifications/src/panel/state/create-listener-middleware.ts]: I just tried this and TypeScript still complained 😅. This seems to be a design limitation of Typescript and there's no perfect solution yet. (see https://github.com/microsoft/TypeScript/issues/9998).

I think it's fine to use `!` to silence the warnings here if we make sure it's always set.

File: apps/notifications/src/panel/Notifications.jsx
Comment (kevin940726) [reply apps/notifications/src/panel/Notifications.jsx:83]: Is there any use case where we want two different `<Notifications>` instances to be rendered simultaneously? I think that's never supported and will break when we do so. That's why I added an additional check in this PR to avoid rendering the global notifications panel when visiting the `/read/notifications`.

> A good solution would be to stop setting the local store, and just import `store` from `../panel/state`.

I think a better solution might be the other way around. Since `connect` and `useSelector` always use the local store from the `<Provider>`. However, I think either way works if we ensure that there's only going to be one store at a time.

Comment (jsnajdr) [reply apps/notifications/src/panel/Notifications.jsx:83]: > Is there any use case where we want two different <Notifications> instances to be rendered simultaneously?

There was the `/read/notifications` page, but that's now fixed. But what remains is that two different `<Notifications>` instances, not simultaneously but one after another, will use the same Redux store. And they need to work together well to prevent bugs like the `n` key not working.

> I think a better solution might be the other way around. Since `connect` and `useSelector` always use the local store from the `<Provider>`.

Yes, I agree this is the final state to which we should eventually refactor. Two things that currently prevent this:
- in the `standalone` app, the `store` is used _above_ the provider, via a global variable
- this global variable is imported at ~5 places (`import { store } from '../state'`). All these usages would need to be refactored to `react-redux` APIs that read the `store` from context

If you want to do this in this PR, I'm all for it. Otherwise the `standalone` app needs to be fixed some other way.

File: apps/notifications/src/panel/state/create-listener-middleware.ts
Comment (jsnajdr) [reply apps/notifications/src/panel/state/create-listener-middleware.ts]: That's interesting because _my_ TypeScript doesn't complain and is happy 🙂 

Screenshot 2024-08-26 at 11 22 41

Comment (kevin940726) [reply apps/notifications/src/panel/state/create-listener-middleware.ts]: [Here's](https://www.typescriptlang.org/play/?#code/C4TwDgpgBAEghgOwCYBsICcoF4oAoCU2AfFAG4D2AlkgNwBQAxuQgM7BQqVsQIYvZReAdygBZOGAA8bdJQQBzADRQAyhGCT4yNOiJEC9Jq3ahIAgOQAVAJoAFAKLnDzNlAAWiVHwBcsTzoBtAF0BYPo6NHZObl50FgAxcnRLcGgcaOAePgA6eXVcKFNofHpKADM8KABCKA4uTNiEpJTIQgBvOgBIDKy4xOTUgWFVdUk-bQwofRKunsbslnzC1MU5vn6W4voAXzoPCbjssqT7OAY3AoL9r0xCLBI1vubU7LgkJCv-ScISoA) the playground for the latest TypeScript version. Could you verify that this solution works on that too?

File: apps/notifications/src/panel/Notifications.jsx
Comment (kevin940726) [reply apps/notifications/src/panel/Notifications.jsx:83]: > But what remains is that two different `<Notifications>` instances, not simultaneously but one after another, will use the same Redux store. And they need to work together well to prevent bugs like the `n` key not working.

Could you elaborate how that won't work with reproduction steps? I think that's what `removeListeners` fixed when the component unmounts.

> If you want to do this in this PR, I'm all for it.

I want to avoid that in this PR and focus on fixing the issue first. 😅

Comment (jsnajdr) [reply apps/notifications/src/panel/Notifications.jsx:83]: > Could you elaborate how that won't work with reproduction steps?

`removeListeners` fixes the popup inside Calypso and its conflict with `/read/notifications`. But I believe that it will be the `standalone` app that will break, namely when receiving messages from the parent window. I can prepare specific reproduction steps, I'm curious about that myself 🙂

File: apps/notifications/src/panel/state/create-listener-middleware.ts
Comment (jsnajdr) [reply apps/notifications/src/panel/state/create-listener-middleware.ts]: Thanks, I see the error in the playground. What is special about the playground is that the `listenersForType` is a global top-level variable, and that it's referenced inside a callback function. Do any of these two things to fix it:
1. wrap the code in a `function foo() { ... }` so that `listenersForType` becomes a local variable. Error disappears.
2. Rewrite the loop to `for ( const handler of handlers )`. Error also disappears.

I think TypeScript covers the case when the callback is called async, and the possibility that the variable is later reassigned. Consider this:
```ts
let listeners = map.get( type ); // Set | undefined
if ( ! listeners ) { listeners = new Set() }; // just Set after this
handlers.then( ( handler ) => listeners.add( handler ) ); // The same code as .forEach but clearly async
/* some intermediate code */
listeners = undefined; // .then handler will fail if it runs after this
```
Here, because `listeners` is not scoped to a small block where all assignments are under control, TypeScript assumes that `listeners = undefined` can happen.

Comment (kevin940726) [reply apps/notifications/src/panel/state/create-listener-middleware.ts]: That fixed it for me! Thanks for the detailed explanation! 🙇 

> Here, because `listeners` is not scoped to a small block where all assignments are under control, TypeScript assumes that `listeners = undefined` can happen.

I thought that the playground automatically assumed ESM mode, but apparently not 😆.

File: apps/notifications/src/panel/Notifications.jsx
Comment (jsnajdr) [reply apps/notifications/src/panel/Notifications.jsx:83]: > I can prepare specific reproduction steps

The code is so convoluted that I had to give up reproducing something 😅 Two tiny things that are maybe broken are:
- on closing the notification panel, selected notification should be reset: if you reopen, nothing should be selected
- when the panel closes or the browser window becomes invisible (`document.visibilityState`), the notifications shouldn't be refreshed in background

But both patched and unpatched versions work the same way, maybe these small features are already broken? I didn't find anything that should block this PR.

**Review by jsnajdr — APPROVED** [review]
:shipit:

**Review by dmsnell — COMMENTED** [review]
Thanks for the ping. I wish I had had the time to give this a proper review before merging, but it seems like you and @jsnajdr covered things pretty well.

This app is full of legacy code, having been built before Redux was built. The main thing I would have wanted to review is the interaction between the embedded and standalone clients. The external store interface presented to Calypso was there to make it possible to hook into the app from the outside while not coupling the inner app to any particular external app. Given that it seems there have been no major regressions, it looks like you did a proper job ensuring this.

**Pull request merged**
